### ISSUE TITLE
Missing symlinks for libblas.so and liblapack.so

### ISSUE THREAD

**Issue by SylvainCorlay** [conda-forge/blas-feedstock#29]
The current version of the openblas package does not include the `.so` symlinks. Only `libblas.so.3` and `liblapack.so.3` are available.

**Comment by isuruf** [comment]
I'm not so sure about this anymore. I've seen a couple of projects (scipy and another) accidentally linking to `libblas.so` in conda-build because of missing `libblas` and end up linking to `libopenblas.so` because of the soname.

How about a `blas-devel` package with the symlinks?

**Comment by jakirkham** [comment]
Are there any cases where the symlinks would be needed in the installation environment?

**Comment by isuruf** [comment]
@jakirkham, only when a package would dlopen the library at runtime.

**Comment by isuruf** [comment]
@SylvainCorlay, can you add a dependency on `blas` for `xtensor-blas`?

**Comment by isuruf** [comment]
ping @SylvainCorlay

**Comment by isuruf** [comment]
@SylvainCorlay, since you didn't comment, I'll assume that you are okay with moving the symlink to `blas` package.

**Comment by minrk** [comment]
Then I would argue that creation of the `libblas-devel` package should be a *prerequisite* for removing the symlinks from `libblas`. Right now, the only solution I see is for packages that do runtime compilation is to depend on `blas` as a runtime dep, so I'm going to go ahead and do that for now as it's the only way to create a working package.

**Comment by minrk** [comment]
To be consistent, `libblas_netlib` should also remove `libblas.so`, otherwise it doesn't really make sense.

**Comment by isuruf** [comment]
PRs welcome. This should also include the header files as well.
There are other options as well.

1. vendor .so files from netlib into the other packages
2. Do what debian does and create a stub library with `libblas.so.3` as the soname.

>Right now, the only solution I see is for packages that do runtime compilation is to depend on blas as a runtime dep, so I'm going to go ahead and do that for now as it's the only way to create a working package.

Sure. Note that you'll have to be careful in downstream packages about not linking to the shared library from an implementation (instead of netlib ones)

**Comment by minrk** [comment]
> PRs welcome

Will do when I get a chance. I'd start with reverting the symlinks back into libblas so things are consistent again while we get libblas-devel set up, if that's okay.

> Note that you'll have to be careful in downstream packages about not linking to the shared library from an implementation (instead of netlib ones)

Yeah, thanks. Not a concern for the packages I'm currently looking at because these are *runtime* compilation, not build stages for redistribution. Failure to compile with `libblas_openblas` is indeed a bug in the package for these cases.

**Comment by minrk** [comment]
Fixed by #39, thanks @isuruf!

**Comment by isuruf** [comment]
@minrk, @SylvainCorlay, have you added a requirement of `blas-devel` to the packages yet? It's been more than a year and I think it's time to revert #39

**Comment by isuruf** [comment]
Ping again

**Comment by minrk** [comment]
I had to pull back from a lot of conda-forge stuff for a while, so I just forgot about it. That's on me, though and https://github.com/conda-forge/petsc-feedstock/pull/106 should do it, since all the packages I deal with ultimately depend on petsc.

**Comment by minrk** [comment]
Hm, there appear to be some hiccups.

`blas-devel` should be in the `host` dependencies of every recipe that links libblas, right? Since there shouldn't be libblas.so in *any* libblas package. Even numpy hasn't been updated to use blas-devel in the host dependencies.



I also see something weird in priority resolution:

```
conda create -n petsc petsc=3.14.5
```

requests `libblas-3.9.0-8_openblas`

However, adding blas-devel switches the default to mkl:

```
conda create -n petsc petsc=3.14.5 blas-devel
```

gets `libblas-3.9.0-8_mkl`

Similarly, if I try to get blas-devel with openblas:

```
conda install openblas blas-devel
```

I get libblas_mkl *and* openblas.

Additionally, `conda install blas-devel=*=*openblas` to explicitly request blas-devel with openblas fails in the first solve:

```
$ conda create -n blasdev blas-devel=*=*_openblas
Collecting package metadata (current_repodata.json): done
Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.
Collecting package metadata (repodata.json): done
Solving environment: done
```

**Comment by minrk** [comment]
It looks like blas-devel package is still incomplete. Tasks left to do before blas-devel can be adopted:

- [ ] add missing netlib variant of `blas-devel`
- [ ] remove unversioned lib*.so from `lib*` netlib variants

Otherwise, blas-devel cannot be used because it conflicts with the `libblas=*=netlib` used in all conda-forge builds:

```
$ conda create -n blas libblas=*=*netlib blas-devel
Collecting package metadata (current_repodata.json): done
Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.
Collecting package metadata (repodata.json): done
Solving environment: |
Found conflicts! Looking for incompatible packages.
This can take several minutes.  Press CTRL-C to abort.
failed

UnsatisfiableError: The following specifications were found to be incompatible with each other:

Output in format: Requested package -> Available versions

Package libblas conflicts for:
blas-devel -> libblas[version='3.8.0|3.9.0',build='0_h43114d0_netlib|7_openblas|7_blis|8_mkl|8_openblas|8_blis|7_mkl|7_hd44dcd8_netlib']
libblas[build=*netlib]
blas-devel -> liblapack==3.9.0[build=*netlib] -> libblas[version='3.8.0.*|3.8.0|3.8.0|3.8.0|3.8.0|3.8.0|3.8.0|3.8.0|3.8.0|3.8.0|3.8.0|3.8.0|3.8.0|3.8.0|3.8.0|3.8.0|3.8.0|3.8.0|3.8.0|3.8.0|3.8.0|3.8.0|3.9.0|3.9.0|3.9.0|3.9.0.*|3.9.0|3.9.0|3.9.0|3.9.0|3.9.0',build='7_h4bb4525_netlib|15_blis|16_openblas|17_openblas|18_mkl|21_mkl|0_h508aa58_netlib|2_blis|3_blis|4_blis|5_openblas|6_blis|6_mkl|6_openblas|5_mkl|5_blis|4_mkl|4_openblas|3_openblas|2_openblas|20_mkl|19_mkl|16_mkl|16_blis|15_mkl|15_openblas|14_mkl|14_openblas|14_blis']
Note that strict channel priority may have removed packages required for satisfiability.

**Comment by minrk** [comment]
Hm, looks like it's there but out-of-date due to the build string: https://github.com/conda-forge/lapack-feedstock/pull/50

**Comment by dalcinl** [comment]
I'm wondering whether conda-forge should slowly move to use [FlexiBlas](https://www.mpi-magdeburg.mpg.de/projects/flexiblas), just as the Fedora community recently did [[link]](https://lists.fedorahosted.org/archives/list/devel@lists.fedoraproject.org/thread/7MT6QJA4A7SG4UCV2HVEUYIRCET2PGHS/#SOMBJFUSN6JK3QGMURA35UEBBFJAMBK4).

**Comment by minrk** [comment]
conda-forge effectively already has that behavior - we build and link against `libblas` (etc.) provided by netlib during build, then at runtime you can select any blas/lapack provider - openblas, netlib, mkl, etc. We haven't had a blas-provider build matrix for some time, thanks to the wonderful work of @isuruf.

This issue is just about exactly what package has `libblas.so` vs `libblas.so.3`.

**Comment by dalcinl** [comment]
@minrk What do you mean by runtime? `conda install` time? Or can users really switch the implementation via config files of environ vars?

**Comment by SylvainCorlay** [comment]
We had been working around that issue in the xtensor-blas binder by simply requiring  `blas * *openblas"` for now. I think making blas-devel a dependency of xtensor-blas is probably a better solution.

Thank you for working this out.

**Comment by minrk** [comment]
@dalcinl yes, `conda install` time. The blas providers are mutually exclusive, so cannot be in the same env.
